src/fastfeedparser/main.py

```diff
@@ -44,640 +47,571 @@ def parse(source):
-            elif content_encoding == 'deflate':
+            elif content_encoding == "deflate":
                 content = zlib.decompress(content, -zlib.MAX_WBITS)
-            xml_content = content
+            content_charset = response.headers.get_content_charset()
+            xml_content = (
+                content.decode(content_charset) if content_charset else content
+            )
     else:
         xml_content = source
 
+    # Ensure we have bytes for lxml
+    if isinstance(xml_content, str):
+        xml_content = xml_content.encode("utf-8", errors="replace")
+
     # Handle empty content
-    if isinstance(xml_content, str) and not xml_content.strip():
+    if not xml_content.strip():
         raise ValueError("Empty content")
 
-    # Handle decoding if content is bytes
-    if isinstance(xml_content, bytes):
-        encodings = ["utf-8", "iso-8859-1", "windows-1252"]
-        decoded = None
-        for encoding in encodings:
-            try:
-                decoded = xml_content.decode(encoding)
-                break
-            except UnicodeDecodeError:
-                continue
-        if decoded is None:
-            raise ValueError("Could not decode content with any supported encoding")
-        xml_content = decoded
-
-    # Ensure we have bytes for lxml
-    xml_content = xml_content.encode("utf-8", errors="replace")
-
-    parser = etree.XMLParser(recover=True)
+    parser = etree.XMLParser(
+        ns_clean=True,
+        recover=True,
+        collect_ids=False,
+        resolve_entities=False,
+    )
     try:
         root = etree.fromstring(xml_content, parser=parser)
     except etree.XMLSyntaxError as e:
         raise ValueError(f"Failed to parse XML content: {str(e)}")
-
-    # Check if root is None
     if root is None:
         raise ValueError("Failed to parse XML content: root element is None")
 
-    namespaces = root.nsmap
-
-    # Clean up the XML tree
-    for element in root.iter():
-        if element.text:
-            element.text = element.text.replace("\x00", "")  # Remove null characters
-        if element.tail:
-            element.tail = element.tail.replace("\x00", "")  # Remove null characters
-
-    feed = FastFeedParserDict()
-    entries = []
-
-    # Determine feed type based on content structure
-    if root.tag == "rss" or root.tag == f"{{{namespaces.get(None, '')}}}rss":
+    # Determine a feed type based on the content structure
+    feed_type: _FeedType
+    if root.tag == "rss" or root.tag.endswith("}rss"):
         feed_type = "rss"
         channel = root.find("channel")
         if channel is None:
             raise ValueError("Invalid RSS feed: missing channel element")
         items = channel.findall("item")
-    elif (
-        root.tag.endswith("feed")
-        or root.tag
-        == f"{{{namespaces.get('atom', 'http://www.w3.org/2005/Atom')}}}feed"
-    ):
+    elif root.tag == "{http://www.w3.org/2005/Atom}feed":
```

Comment:
BREAKING? Here I made Atom tag checking more strict because if it doesn't crash here it will later, we rely heavily on the 2005/Atom schema. I don't think it's okay to accept any schema here but I could be wrong.